pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/NullableSingleInputAggregationFunction.java

```diff
@@ -0,0 +1,263 @@
+   */
+  public void forEachNotNull(int length, BlockValSet blockValSet, BatchConsumer consumer) {
+    if (!_nullHandlingEnabled) {
+      consumer.consume(0, length);
+      return;
+    }
+
+    RoaringBitmap roaringBitmap = blockValSet.getNullBitmap();
+    if (roaringBitmap == null) {
+      consumer.consume(0, length);
+      return;
+    }
+
+    forEachNotNull(length, roaringBitmap.getIntIterator(), consumer);
+  }
+
+  /**
+   * Iterates over the non-null ranges of the nullIndexIterator and calls the consumer for each range.
+   * @param nullIndexIterator an int iterator that returns values in ascending order whose min value is 0.
+   *                          Rows are considered null if and only if their index is emitted.
+   */
+  public void forEachNotNull(int length, IntIterator nullIndexIterator, BatchConsumer consumer) {
+    int prev = 0;
+    while (nullIndexIterator.hasNext() && prev < length) {
+      int nextNull = Math.min(nullIndexIterator.next(), length);
+      if (nextNull > prev) {
+        consumer.consume(prev, nextNull);
+      }
+      prev = nextNull + 1;
+    }
+    if (prev < length) {
+      consumer.consume(prev, length);
+    }
+  }
+
+  /**
+   * Folds over the non-null ranges of the blockValSet using the reducer.
+   * @param initialAcum the initial value of the accumulator
+   * @param <A> The type of the accumulator
+   */
+  public <A> A foldNotNull(int length, @Nullable RoaringBitmap roaringBitmap, A initialAcum, Reducer<A> reducer) {
+    IntIterator intIterator = roaringBitmap == null ? null : roaringBitmap.getIntIterator();
+    return foldNotNull(length, intIterator, initialAcum, reducer);
+  }
+
+  /**
+   * Folds over the non-null ranges of the nullIndexIterator using the reducer.
+   * @param nullIndexIterator an int iterator that returns values in ascending order whose min value is 0.
+   *                          Rows are considered null if and only if their index is emitted.
+   * @param initialAcum the initial value of the accumulator
+   * @param <A> The type of the accumulator
+   */
+  public <A> A foldNotNull(int length, @Nullable IntIterator nullIndexIterator, A initialAcum, Reducer<A> reducer) {
+    A acum = initialAcum;
+    if (!_nullHandlingEnabled || nullIndexIterator == null || !nullIndexIterator.hasNext()) {
+      return reducer.apply(initialAcum, 0, length);
+    }
+
+    int prev = 0;
+    while (nullIndexIterator.hasNext() && prev < length) {
+      int nextNull = Math.min(nullIndexIterator.next(), length);
+      if (nextNull > prev) {
+        acum = reducer.apply(acum, prev, nextNull);
+      }
+      prev = nextNull + 1;
+    }
+    if (prev < length) {
+      acum = reducer.apply(acum, prev, length);
+    }
+    return acum;
+  }
+
+  /**
+   * Returns an iterator that returns the union of the null bitmaps of valSet1 and valSet2.
+   */
+  public IntIterator orNullIterator(BlockValSet valSet1, BlockValSet valSet2) {
```

Comment:
Is this used? Trying to understand where this method fits

Reply:
It is not used here, but it is being used in other PRs. Given I have so many PRs sharing the same base code, I've been merging them whenever I found an issue in the other PRs. For example, this method is being used in https://github.com/apache/pinot/pull/12271.

I guess I can remove it from here if I carefully merge them later